compiler/qsc_passes/src/capabilitiesck.rs

```diff
@@ -0,0 +1,538 @@
+    UseOfDynamicBigInt(#[label] Span),
+
+    #[error("cannot use a dynamic string value")]
+    #[diagnostic(help(
+        "using a dynamic string value, a string value that depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicString"))]
+    UseOfDynamicString(#[label] Span),
+
+    #[error("cannot use a dynamically-sized array")]
+    #[diagnostic(help(
+        "using a dynamically-sized array, an array whose size depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicallySizedArray"))]
+    UseOfDynamicallySizedArray(#[label] Span),
+
+    #[error("cannot use a dynamic user-defined type")]
+    #[diagnostic(help(
+        "using a dynamic user-defined type, a user-defined type in which one or more of its members depend on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicUdt"))]
+    UseOfDynamicUdt(#[label] Span),
+
+    #[error("cannot use a dynamic function")]
+    #[diagnostic(help(
+        "using a dynamically resolved function, a function whose resolution depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicArrowFunction"))]
+    UseOfDynamicArrowFunction(#[label] Span),
+
+    #[error("cannot use a dynamic operation")]
+    #[diagnostic(help(
+        "using a dynamically resolved operation, an operation whose resolution depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.UseOfDynamicArrowOperation"))]
+    UseOfDynamicArrowOperation(#[label] Span),
+
+    #[error("cannot call a cyclic function with a dynamic value as argument")]
+    #[diagnostic(help(
+        "calling a cyclic function with a dynamic value as argument, an argument value that depends on a measurement result, is not supported by the target"
+    ))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.CallToCyclicFunctionWithDynamicArg"))]
+    CallToCyclicFunctionWithDynamicArg(#[label] Span),
+
+    #[error("cannot define a cyclic operation specialization")]
+    #[diagnostic(help("cyclic operation specializations, specializations that contain cycles, are not supported by the target"))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.CyclicOperationSpec"))]
+    CyclicOperationSpec(#[label] Span),
+
+    #[error("cannot call a cyclic operation")]
+    #[diagnostic(help("calling a cyclic operation is not supported by the target"))]
+    #[diagnostic(code("Qsc.CapabilitiesCk.CallToCyclicOperation"))]
+    CallToCyclicOperation(#[label] Span),
+
+    #[error("cannot call a function or operation whose resolution is dynamic")]
+    #[diagnostic(help("calling a function or operation whose resolution is dynamic, a resolution that depends on a measurement result, is not supported by the target"))]
```

Comment:
nitpick: I might be worth thinking a bit about whether we can find a different term other than "resolution" because that can lead to some confusion. This came up when working on the compiler internals that use the term, and it is the right noun to use, but might not be the clearest thing for a help message. That said, I don't have any alternative suggestions, so feel free to table this one for a later discussion with a broader group.

Reply:
I included this in #1285, which tracks the names/words we want to re-consider.